File: 3319-k-th-largest-perfect-subtree-size-in-binary-tree/3319-k-th-largest-perfect-subtree-size-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int dfs(TreeNode* root,vector<int> &v){
        if(root==NULL){
            return 0;
        }
        int l=dfs(root->left,v);
        int r=dfs(root->right,v);

        if(l==r){
            v.push_back(l+r+1);
            return l+r+1;
        }
        return -1;
    }
    int kthLargestPerfectSubtree(TreeNode* root, int k) {
        vector<int> v;
        int z=dfs(root,v);
        sort(v.rbegin(),v.rend());
        return k>v.size()?-1:v[k-1];
    }
};

File: 3319-k-th-largest-perfect-subtree-size-in-binary-tree/3319-k-th-largest-perfect-subtree-size-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // To store depths of perfect subtrees
    List<Integer> a = new ArrayList<>();

    public int dfs(TreeNode root) {
        if (root == null) {
            return 0;
        }

        int left = dfs(root.left);
        int right = dfs(root.right);

        if (left == -1 || right == -1 || left != right) {
            return -1;
        }
        a.add(left + 1);
        return left + 1;
    }

    public int kthLargestPerfectSubtree(TreeNode root, int k) {
        dfs(root);
        Collections.sort(a, Collections.reverseOrder());
        if (a.size() < k) {
            return -1;
        }
        return (1 << a.get(k - 1)) - 1;
    }
}